File: system-test/dotnet/Driver.Adapter/MyShop/Api/Client/Controllers/CouponController.cs
using Common;
using Driver.Adapter.Shared.Client.Http;
using Driver.Adapter.MyShop.Api.Client.Dtos.Errors;
using Driver.Port.MyShop.Dtos;

namespace Driver.Adapter.MyShop.Api.Client.Controllers;

public class CouponController
{
    private const string Endpoint = "/api/coupons";

    private readonly JsonHttpClient<ProblemDetailResponse> _httpClient;

    public CouponController(JsonHttpClient<ProblemDetailResponse> httpClient)
    {
        _httpClient = httpClient;
    }

    public Task<Result<VoidValue, ProblemDetailResponse>> PublishCouponAsync(PublishCouponRequest request)
        => _httpClient.PostAsync(Endpoint, request);

    public Task<Result<BrowseCouponsResponse, ProblemDetailResponse>> BrowseCouponsAsync()
        => _httpClient.GetAsync<BrowseCouponsResponse>(Endpoint);
}

File: system-test/java/src/main/java/com/mycompany/myshop/testkit/driver/adapter/myshop/api/client/controllers/CouponController.java
package com.mycompany.myshop.testkit.driver.adapter.myshop.api.client.controllers;

import com.mycompany.myshop.testkit.driver.adapter.myshop.api.client.dtos.errors.ProblemDetailResponse;
import com.mycompany.myshop.testkit.driver.adapter.shared.client.http.JsonHttpClient;
import com.mycompany.myshop.testkit.driver.port.myshop.dtos.BrowseCouponsResponse;
import com.mycompany.myshop.testkit.driver.port.myshop.dtos.PublishCouponRequest;
import com.mycompany.myshop.testkit.common.Result;

public class CouponController {
    private static final String ENDPOINT = "/api/coupons";

    private final JsonHttpClient<ProblemDetailResponse> httpClient;

    public CouponController(JsonHttpClient<ProblemDetailResponse> httpClient) {
        this.httpClient = httpClient;
    }

    public Result<Void, ProblemDetailResponse> publishCoupon(PublishCouponRequest request) {
        return httpClient.post(ENDPOINT, request);
    }

    public Result<BrowseCouponsResponse, ProblemDetailResponse> browseCoupons() {
        return httpClient.get(ENDPOINT, BrowseCouponsResponse.class);
    }
}
